Add Arrowheads to Arrows

steveruizok opened the pull request:
This PR adds start and end decorators to arrows.

![image](https://user-images.githubusercontent.com/23072548/101247415-6d244f80-3711-11eb-8351-f664ee582352.png)

A user can add decorators to arrow elements.

Each arrow element can have a **start** and **end** decorator.

A decorator may be **null**, **arrow**, **dot**, or **bar**.

By default, an arrow uses only an end decorator of **arrow**.

![Kapture 2020-12-05 at 15 54 16](https://user-images.githubusercontent.com/23072548/101247564-2edb6000-3712-11eb-8abc-d2d54b0516e7.gif)

## Notes

- I've implemented a strategy to correctly render arrows without decorators, such as those from earlier projects; however I'm sure there is an existing approach to handle this sort of thing. I appreciate any notes or guidance here!
- This PR also includes small changes to the icons in the panel.

dwelle commented:
How about a dropdown instead of cycling?

lipis commented:
For dropdown we have to make a custom element to support icons and it will be pain to make it for mobile.. can we have the first iteration of this feature in that form?

lipis commented:
Also for the first version I would just support the arrow and not the rest, and in a separate PR choose which endings we want to add as a start.

dwelle commented:
> Also for the first version I would just support the arrow and not the rest, and in a separate PR choose which endings we want to add as a start.

I think we can leave it if it's implemented already, at least it will force us to implement the dropdown sooner.

One thing that bothers me a bit is that when you remove caps and then create new arrows it remembers the setting and effectively makes new arrows into lines. I'm not sure whether we should do anything about it though.

steveruizok commented:
Thanks for the feedback.

> Also for the first version I would just support the arrow and not the rest, and in a separate PR choose which endings we want to add as a start.

@lipis I've gone ahead and removed the other two decorators as you suggested. This PR now only proposes the ability to add or remove arrow decorators to the start and end of an arrow.

I believe this approach was also suggested in the other discussion. However, given the uncertainty around the selection UI, I wanted to have more than one option: when there are only two options, the buttons function just like a checkbox, so we really needed 3+ to judge the cycle interaction. Let's refer back to [this preview](https://excalidraw-git-arrow-options.excalidraw.vercel.app) deploy when discussing that interaction.

@dwelle As for the cycle interaction itself, I'm not 100% sure on whether it's the right approach. Most design tools have a panel to select line caps / decorators. I don't think a graphical select would be the right choice—hard to implement, overkill for few decorators and wouldn't scale easily to many decorators.

As an alternative to the cycle button, we could instead use a native text-based select but still show the icon of the selected decorator.

![image](https://user-images.githubusercontent.com/23072548/101259202-5be23f80-371f-11eb-937c-727160eb07b2.png)

That said, I this a cycle button like the one implemented [here](https://excalidraw-git-arrow-options.excalidraw.vercel.app) is a fine choice for maybe three or four cap styles. And it works for one style, too, as implemented [here](https://excalidraw-32fnzbnku.vercel.app/)—as mentioned above, it works just like a checkbox with only two options.

lipis reviewed (approved):
I like it for initial release is perfect! For the upcoming releases we can discuss different options on how to scale and with which endings.

lipis commented on src/locales/en.json:
```diff
@@ -28,6 +28,11 @@
+    "nullDecorator": "None",
+    "arrowDecorator": "Arrow",
+    "dotDecorator": "Dot",
+    "barDecorator": "Bar",
```
let's remove the unused ones

lipis commented on src/components/icons.tsx:
```diff
@@ -733,11 +720,75 @@ export const EdgeRoundIcon = React.memo(
+    ),
+);
+
+export const BarDecoratorIcon = React.memo(
```
let's remove the unused icons in this PR

steveruizok commented:
Could you check up on how I'm handling older arrows without start / end decorators? I'm unsure whether that part of this PR is in line with the rest of the project.

lipis commented on src/components/icons.tsx:
```diff
@@ -733,11 +720,75 @@ export const EdgeRoundIcon = React.memo(
+    ),
+);
+
+export const DotDecoratorIcon = React.memo(
```
remove this one as well

lipis reviewed (changes requested):
Some left overs

dwelle commented:
> Could you check up on how I'm handling older arrows without start / end decorators? I'm unsure whether that part of this PR is in line with the rest of the project.

I'll do a review of the code, but no sooner than tomorrow.

lipis commented:
Run `npm run test:update`

lipis reviewed (changes requested):
When two arrows and one rectangle are selected, the decorators don't show the actual status.

steveruizok commented:
> When two arrows and one rectangle are selected, the decorators don't show the actual status.

Good catch, fixing in the next commit. Tests are also updated and passing.

One of the tests in move.test.tsx, `rectangles with binding arrow`, has the UI creating a line element instead of an arrow. I fixed this in the test, but was this intended?

lipis commented:
> One of the tests in move.test.tsx, `rectangles with binding arrow`, has the UI creating a line element instead of an arrow. I fixed this in the test, but was this intended?

Let's wait for @dwelle on this one

steveruizok commented:
Strangely that same test (`rectangles with binding arrow`) passes locally but fails when run by the Github action. :shrug:

lipis commented:
Robots are failing us 🤷 🤖

dwelle commented:
> One of the tests in move.test.tsx, `rectangles with binding arrow`, has the UI creating a line element instead of an arrow. I fixed this in the test, but was this intended?

I haven't written that one, but ultimately it doesn't matter whether it's a line or an arrow because in the test we're binding it via an API which is agnostic (works with either).

steveruizok commented:
> ...ultimately it doesn't matter whether it's a line or an arrow because in the test we're binding it via an API which is agnostic (works with either).

Makes sense. The test only fails because it expects the "arrow" to have a "arrow" end decorator, but that decorator is only applied to new arrow elements, not to lines. Unless were explicitly testing whether binding also works on lines, I'd just make it an arrow.

lipis reviewed (approved):
LGTM

lipis commented on src/renderer/renderElement.ts:
```diff
@@ -334,22 +336,64 @@ const generateElementShape = (
-          if (arrowPoints) {
-            const [x2, y2, x3, y3, x4, y4] = arrowPoints;
-            // for dotted arrows caps, reduce gap to make it more legible
+          const { startDecorator = null, endDecorator = "arrow" } = element;
```
What about using enum or some sort of type for the decorators?

(Maybe we should also use that for shapes and other strings.. but that's for a separate PR)

dwelle replied:
Not sure what you mean, these are already typed.

Note: I'd rather not use enums in general (though I'm aware we use them in one or two places already), as it's a TS feature that has many weird rules that are hard to remember. It's also introduces runtime constructs (i.e. it's not erasable so we won't ever be able to use tools which only erase types, such as esbuild). And lastly, its use is discouraged by TS maintainers themselves.

lipis replied:
`DECORATOR.ARROW` instead of strings.. for autocompletion and everything

dwelle replied:
They're already typed so there's no need for adding constants on top of them, we'd end up having constants for every typed string, which would be duplicating of the effort.

Note: I'd rather not use enums in general (though I'm aware we use them in one or two places already), as it's a TS feature that has many weird rules that are hard to remember. It's also introduces runtime constructs (i.e. it's not erasable so we won't ever be able to use tools which only erase types, such as esbuild). And lastly, its use is discouraged by TS maintainers themselves.

steveruizok replied:
> What about using enum or some sort of type for the decorators?

For what it's worth, ESLint is also not set up to handle enums. (It throws an unused-var warning).

dwelle commented on src/locales/en.json:
```diff
@@ -28,6 +28,9 @@
     "edges": "Edges",
     "sharp": "Sharp",
     "round": "Round",
+    "decorators": "Decorators",
```
Is "decorators" a term used in the wild? I've seen endings or caps.

lipis replied:
if we change the label I think we should also change the internal variable everywhere as well..

I vote for `endings` if not `decorators`

steveruizok replied:
I'm not sure where I picked it up from, possibly leaflet (https://github.com/bbecquet/Leaflet.PolylineDecorator). I'm open to alternatives though. Cap and join usually refer to the stroke shape, rather than some separate shape or symbol that uh decorates the start or end of the line. Different apps call them different names no absolute convention.

Sketch calls them "arrowheads". Of all the options, this feels the most excalidrawy to me.

lipis replied:
`arrow heads` is two words.. `caps` doesn't feel right and confusing, `endings` is straight forward.. and `decorators` is not bad :)

steveruizok replied:
> `arrow heads` is two words

Just one actually! (sorry 👨🏻‍🏫)

But given that we're adding shapes to both heads and tails, arrowheads might not be a good fit anyway. That's my only issue with endings, too: we'd have `endings.start` and `endings.end`. There's also the distant possibility of adding shapes to joins or in a rhythm along a line, which makes sense for some diagrams (see the Leaflet library above).

"Decorators" makes the most sense to me on an API level, because it leaves room for all of that, but it's an awful label in the UI. Endings or Arrowheads are better for the label. If we want to use the same term in both API and label then I would be happy with either term.

lipis replied:
> Just one actually! (sorry 👨🏻‍🏫)

On the properties window it will be called what? `Arrowhead` or `Arrow head`?

lipis replied:
You have a good point about keeping internally the term `decorator`, and I don't really mind having it decorators as a label as well. It's not the worse :)

steveruizok replied:
Yep, **Arrowheads**.

![image](https://user-images.githubusercontent.com/23072548/101292789-91635780-3809-11eb-94cb-8ac9fe3d1af4.png)

lipis replied:
head and tail if we want to be precise :)

dwelle commented on src/scene/comparisons.ts:
```diff
@@ -28,6 +28,8 @@ export const canChangeSharpness = (type: string) =>
 
 export const hasText = (type: string) => type === "text";
 
+export const hasDecorators = (type: string) => type === "arrow";
```
nit: not sure if this shouldn't be called `canHaveDecorators` as `hasDecorates` will suggest it has non-null decorators.

steveruizok replied:
Makes sense to me 👍

lipis commented on src/element/types.ts:
```diff
@@ -90,13 +90,17 @@ export type PointBinding = {
 export type ExcalidrawLinearElement = _ExcalidrawElementBase &
   Readonly<{
-    type: "arrow" | "line" | "draw";
+    type: "line" | "draw" | "arrow";
```
Let's do alphabetical but ASC if anything :)

steveruizok replied:
Honestly no intent there, not sure why I didn't just append it!

lipis reviewed (approved):
Let's keep `Decorators` and ship it :)

steveruizok commented on src/element/newElement.ts:
```diff
@@ -30,6 +31,11 @@ type ElementConstructorOpts = MarkOptional<
   | "versionNonce"
 >;
 
+type LinearElementConstructorOpts = {
+  startDecorator: LinearElementDecorator | null;
+  endDecorator: LinearElementDecorator | null;
+};
```
I'm not sure I follow—or at least, I can't seem to merge these types without getting warnings elsewhere. Could you clarify?

dwelle replied:
I pushed the fix. It just removes the unnecessary type alias because it's not reused anywhere else.

steveruizok replied:
Ah that makes sense, thanks!

lipis commented on .gitignore:
```diff
@@ -21,3 +21,5 @@ dist/
 .env.test.local
 .env.production.local
+
+.vercel
```
While we at it.. can we sort the whole file (VS Code)

steveruizok replied:
What do you mean, just sort these lines alphabetically? The gitignore still has a `now` in there, though that probably doesn't need to be there anymore.

lipis replied:
yes.. we can get rid of now and sort the whole file alphabetically

lipis commented on src/components/icons.tsx:
```diff
@@ -733,11 +716,33 @@ export const EdgeRoundIcon = React.memo(
+    flip = false,
+  }: {
+    appearance: "light" | "dark";
+    flip?: boolean;
```
I like the flip :) ❤️

steveruizok commented on src/components/App.tsx:
```diff
@@ -2892,6 +2892,14 @@ class App extends React.Component<ExcalidrawProps, AppState> {
         pointerDownState.origin.y,
         elementType === "draw" ? null : this.state.gridSize,
       );
+
+      // If arrow is pre-decorators, it will have undefined for both start and end decorators.
+      // If so, we want it to be null/"arrow". If the linear item is not an arrow, we want it
+      // to be null/null. Otherwise, we want it to use the currentItemLinearDecorators values.
+      const { start, end } = this.state.currentItemLinearDecorators;
+      const [startDecorator = null, endDecorator = "arrow"] =
+        elementType === "arrow" ? [start, end] : [null, null];
+
```
I believe this was to catch arrows being copy and pasted between an older pre-decorators version and PR's version. I suppose this would never occur in production, though, so I'll remove this bit.

dwelle commented:
@steveruizok you went a bit too far on my suggestion to remove the "migration" --- we still wanna retain the `[null, null]` for non-arrows (for now) — hence the failing snapshots.

lipis commented on src/actions/actionProperties.tsx:
```diff
@@ -622,3 +626,99 @@ export const actionChangeSharpness = register({
+
+export const actionChangeDecorator = register({
+  name: "changeDecorator",
+  perform: (elements, appState, value) => {
```
can we track it somewhere? Here is the only place that the magic happens.. right?

```
trackEvent(EVENT_CHANGE, "decorator start",  "arrow");
trackEvent(EVENT_CHANGE, "decorator end",  "null");
trackEvent(EVENT_CHANGE, "decorator end",  "arrow");
```

Reference: [`trackEvent`](https://github.com/excalidraw/excalidraw/blob/master/src/analytics.ts#L13-L20)

Example: [`actionProperties.tsx`](https://github.com/excalidraw/excalidraw/blob/master/src/actions/actionProperties.tsx#L127-L133)

dwelle commented:
> I believe this was to catch arrows being copy and pasted between an older pre-decorators version and PR's version. I suppose this would never occur in production, though, so I'll remove this bit.

That's a good point. I thought we're restoring on paste, but apparently not. Will create an issue.

steveruizok commented:
A few updates:

- I've changed all the names (both labels and in the API) to **arrowheads** / **Arrowheads**. The name is used by Adobe, Sketch, and several other apps. A robust user research study (a [Twitter poll](https://twitter.com/steveruizok/status/1335697807781605379)) confirms that the name is preferred by Excalidraw's users.
- I've added **event tracking**. Changing the starting arrowhead will fire `arrowhead start` and changing the end arrowhead will fire `arrowhead end`. Both events will fire values of either `none`, `arrow`, or any of our future arrowhead names.
- Simplified the `ButtonIconCycle` component to be a fully controlled component.
- Minor other issues resolved above.

I think it's good to go!

dwelle commented:
:heart: :rocket:

vinaysharma14 commented:
Is this feature released? I'm unable to see it.

vinaysharma14 commented:
Is this feature released? I'm unable to see it.

<img width="266" alt="Screenshot 2022-01-27 at 11 02 35 PM" src="https://user-images.githubusercontent.com/48080660/151412402-4d9cd280-e617-4e39-a3da-27915edc7af4.png">
